worker/filenotifywatcher/watcher.go

```diff
@@ -0,0 +1,192 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package filenotifywatcher
+
+import (
+	"path/filepath"
+
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	"github.com/juju/worker/v3"
+	"github.com/juju/worker/v3/catacomb"
+	"k8s.io/utils/inotify"
+)
+
+const (
+	defaultWatcherPath = "/var/lib/juju/locks"
+)
+
+// FileWatcher is an interface that allows a worker to watch a file for changes.
+type FileWatcher interface {
+	worker.Worker
+	// Changes returns a channel that will receive a value whenever the
+	// watched file changes.
+	Changes() <-chan bool
+}
+
+// INotifyWatcher is an interface that allows a worker to watch a file for
+// changes using inotify.
+type INotifyWatcher interface {
+	// Watch adds the given file or directory (non-recursively) to the watch.
+	Watch(path string) error
+
+	// Event returns the next event.
+	Events() <-chan *inotify.Event
+
+	// Error returns the next error.
+	Errors() <-chan error
+
+	// Close removes all watches and closes the events channel.
+	Close() error
+}
+
+type option struct {
+	path      string
+	logger    Logger
+	watcherFn func() (INotifyWatcher, error)
+}
+
+type Option func(*option)
+
+// WithPath is an option for NewWatcher that specifies the path to watch.
+func WithPath(path string) Option {
+	return func(o *option) {
+		o.path = path
+	}
+}
+
+// WithLogger is an option for NewWatcher that specifies the logger to use.
+func WithLogger(logger Logger) Option {
+	return func(o *option) {
+		o.logger = logger
+	}
+}
+
+// WithINotifyWatcherFn is an option for NewWatcher that specifies the inotify
+// watcher to use.
+func WithINotifyWatcherFn(watcherFn func() (INotifyWatcher, error)) Option {
+	return func(o *option) {
+		o.watcherFn = watcherFn
+	}
+}
+
+func newOption() *option {
+	return &option{
+		path:      defaultWatcherPath,
+		logger:    loggo.GetLogger("juju.worker.filenotifywatcher"),
+		watcherFn: newWatcher,
+	}
+}
+
+type Watcher struct {
+	catacomb catacomb.Catacomb
+
+	namespace string
+	changes   chan bool
+
+	watchPath string
+	watcher   INotifyWatcher
+
+	logger Logger
+}
+
+func NewWatcher(namespace string, opts ...Option) (FileWatcher, error) {
+	o := newOption()
+	for _, opt := range opts {
+		opt(o)
+	}
+
+	watcher, err := o.watcherFn()
+	if err != nil {
+		return nil, errors.Annotatef(err, "creating watcher for namespace %q", namespace)
+	}
+	if err := watcher.Watch(o.path); err != nil {
+		return nil, errors.Annotatef(err, "watching namespace %q", namespace)
+	}
+
+	w := &Watcher{
+		namespace: namespace,
+		changes:   make(chan bool),
+		watcher:   watcher,
+		watchPath: filepath.Join(o.path, namespace+".db.lock"),
```

Comment:
Make this more general. Forget the "db" segment, and make the caller supply a name-space that uniquely identifies the purpose.